### ISSUE TITLE
Simple hyperparameter and architecture search in tensorflow with ray tune

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#2367]
Simple hyperparameter and architecture search in tensorflow with ray tune<br>
<br>
http://bit.ly/2QsOvzn<br>
<br>
Louis Kirsch<br>
<br>

<div><div>
<p>In a previous <a href="http://louiskirsch.com/ai/population-based-training">blog post</a> I have shown a very clean method on how to implement efficient hyperparameter search in tensorflow from scratch.
I presented <a href="http://louiskirsch.com/ai/population-based-training#population-based-training-pbt">population-based training</a>, an evolutionary method that allows cheap and adaptive hyperparameter search by changing hyperparameters already during training instead of having to train until convergence before the resulting performance statistics can be used to inform the choice of hyperparameters.
That said, the vanilla version I presented had two major downsides: The training was not distributed out of the box and also required that the graph is constructed in the beginning and can be reused for different parameter settings.</p>

<p>In this blog post we want to look at the distributed computation framework <a href="https://github.com/ray-project/ray">ray</a> and its little brother <a href="http://ray.readthedocs.io/en/latest/tune.html">ray tune</a> that allow distributed and easy to implement hyperparameter search.
It not only supports population-based training but also other hyperparameter search algorithms.
Ray and ray tune support any autograd package, including tensorflow and PyTorch.</p>

<figure>
  <img src="http://louiskirsch.com/assets/posts/ray-tune/preview.png" alt="The architecture of ray tune">
  <figcaption>The architecture of ray tune.</figcaption>
</figure>

<h2 id="setting-up-ray">Setting up ray</h2>

<p>Install ray on all your machines</p>

<p>If you only have a single machine at your disposal (it will make use of all local CPU and GPU resources), initialize ray with</p>

<p>Otherwise, pick a machine to be the head node</p>
<div><div><pre><code>ray start <span>--head</span> <span>--redis-port</span><span>=</span>6379
</code></pre></div></div>
<p>and connect your other instances</p>
<div><div><pre><code>ray start <span>--redis-address</span><span>=</span>HEAD_HOSTNAME:6379
</code></pre></div></div>
<p>Finally, run your code on any node</p>
<div><div><pre><code><span>import</span> <span>ray</span>

<span>ray</span><span>.</span><span>init</span><span>(</span><span>redis_address</span><span>=</span><span>'HEAD_HOSTNAME:6379'</span><span>)</span>
</code></pre></div></div>

<p>Having done that, we are ready to distribute tasks such as hyperparameter tuning.
More information on how to set up a cluster can be found <a href="http://ray.readthedocs.io/en/latest/using-ray-on-a-cluster.html">here</a>.</p>

<h2 id="implementing-your-model">Implementing your model</h2>

<p>We will need to implement a model using the following skeleton.
You should be able to plug in your existing tensorflow code easily:</p>
<div><div><pre><code><span>import</span> <span>ray.tune</span> <span>as</span> <span>tune</span>

<span>class</span> <span>Model</span><span>:</span>
  <span># TODO implement</span>

<span>class</span> <span>MyTrainable</span><span>(</span><span>Trainable</span><span>):</span>
  <span>def</span> <span>_setup</span><span>(</span><span>self</span><span>):</span>
    <span># Load your data</span>
    <span>self</span><span>.</span><span>data</span> <span>=</span> <span>...</span>
    <span># Setup your tensorflow model</span>
    <span># Hyperparameters for this trial can be accessed in dictionary self.config</span>
    <span>self</span><span>.</span><span>model</span> <span>=</span> <span>Model</span><span>(</span><span>self</span><span>.</span><span>data</span><span>,</span> <span>hyperparameters</span><span>=</span><span>self</span><span>.</span><span>config</span><span>)</span>
    <span># To save and restore your model</span>
    <span>self</span><span>.</span><span>saver</span> <span>=</span> <span>tf</span><span>.</span><span>train</span><span>.</span><span>Saver</span><span>()</span>
    <span># Start a tensorflow session</span>
    <span>self</span><span>.</span><span>sess</span> <span>=</span> <span>tf</span><span>.</span><span>Session</span><span>()</span>

  <span>def</span> <span>_train</span><span>(</span><span>self</span><span>):</span>
    <span># Run your training op for n iterations</span>
    <span>for</span> <span>_</span> <span>in</span> <span>range</span><span>(</span><span>n</span><span>):</span>
      <span>self</span><span>.</span><span>sess</span><span>.</span><span>run</span><span>(</span><span>self</span><span>.</span><span>model</span><span>.</span><span>training_op</span><span>)</span>

    <span># Report a performance metric to be used in your hyperparameter search</span>
    <span>validation_loss</span> <span>=</span> <span>self</span><span>.</span><span>sess</span><span>.</span><span>run</span><span>(</span><span>self</span><span>.</span><span>model</span><span>.</span><span>validation_loss</span><span>)</span>
    <span>return</span> <span>tune</span><span>.</span><span>TrainingResult</span><span>(</span><span>timesteps_this_iter</span><span>=</span><span>n</span><span>,</span> <span>mean_loss</span><span>=</span><span>validation_loss</span><span>)</span>

  <span>def</span> <span>_stop</span><span>(</span><span>self</span><span>):</span>
    <span>self</span><span>.</span><span>sess</span><span>.</span><span>close</span><span>()</span>

  <span># This function will be called if a population member</span>
  <span># is good enough to be exploited</span>
  <span>def</span> <span>_save</span><span>(</span><span>self</span><span>,</span> <span>checkpoint_dir</span><span>):</span>
    <span>path</span> <span>=</span> <span>checkpoint_dir</span> <span>+</span> <span>'/save'</span>
    <span>return</span> <span>self</span><span>.</span><span>saver</span><span>.</span><span>save</span><span>(</span><span>self</span><span>.</span><span>sess</span><span>,</span> <span>path</span><span>,</span> <span>global_step</span><span>=</span><span>self</span><span>.</span><span>_timesteps_total</span><span>)</span>

  <span># Population members that perform very well will be</span>
  <span># exploited (restored) from their checkpoint</span>
  <span>def</span> <span>_restore</span><span>(</span><span>self</span><span>,</span> <span>checkpoint_path</span><span>):</span>
    <span>return</span> <span>self</span><span>.</span><span>saver</span><span>.</span><span>restore</span><span>(</span><span>self</span><span>.</span><span>sess</span><span>,</span> <span>checkpoint_path</span><span>)</span>
</code></pre></div></div>
<p>A new Trainable will be instantiated and executed on an available GPU in your cluster / on your machine for each trial or population member (each having their own hyperparameters) in your population-based training.</p>

<h2 id="setting-up-ray-tune">Setting up ray tune</h2>

<p>Next, register your trainable and specify your experiments</p>
<div><div><pre><code><span>import</span> <span>numpy</span> <span>as</span> <span>np</span>

<span>tune</span><span>.</span><span>register_trainable</span><span>(</span><span>'MyTrainable'</span><span>,</span> <span>MyTrainable</span><span>)</span>
<span>train_spec</span> <span>=</span> <span>{</span>
  <span>'run'</span><span>:</span> <span>'MyTrainable'</span><span>,</span>
  <span># Specify the number of CPU cores and GPUs each trial requires</span>
  <span>'trial_resources'</span><span>:</span> <span>{</span><span>'cpu'</span><span>:</span> <span>1</span><span>,</span> <span>'gpu'</span><span>:</span> <span>1</span><span>},</span>
  <span>'stop'</span><span>:</span> <span>{</span><span>'timesteps_total'</span><span>:</span> <span>20000</span><span>},</span>
  <span># All your hyperparameters (variable and static ones)</span>
  <span>'config'</span><span>:</span> <span>{</span>
    <span>'batch_size'</span><span>:</span> <span>20</span><span>,</span>
    <span>'units'</span><span>:</span> <span>100</span><span>,</span>
    <span>'l1_scale'</span><span>:</span> <span>lambda</span> <span>cfg</span><span>:</span> <span>return</span> <span>np</span><span>.</span><span>random</span><span>.</span><span>uniform</span><span>(</span><span>1e-3</span><span>,</span> <span>1e-5</span><span>),</span>
    <span>'learning_rate'</span><span>:</span> <span>tune</span><span>.</span><span>random_search</span><span>([</span><span>1e-3</span><span>,</span> <span>1e-4</span><span>])</span>
    <span>...</span>
  <span>},</span>
  <span># Number of trials</span>
  <span>'repeat'</span><span>:</span> <span>4</span>
<span>}</span>
</code></pre></div></div>
<p>The entry <em>‘repeat’</em> describes the number of trials / population members.
Each trial will sample its <em>‘config’</em> from the specification above (i.e. using the predefined values or running the specified function).
The instruction <code>tune.random_search</code> will multiply the number of trials by the number of elements it was given, effectively running a grid search.</p>

<p>Finally, we have to define the kind of hyperparameter tuning we would like to perform and start our experiments.
This is how it works for population-based training:</p>
<div><div><pre><code><span>pbt</span> <span>=</span> <span>PopulationBasedTraining</span><span>(</span>
  <span>time_attr</span><span>=</span><span>'training_iteration'</span><span>,</span>
  <span>reward_attr</span><span>=</span><span>'mean_loss'</span><span>,</span>
  <span>perturbation_interval</span><span>=</span><span>1</span><span>,</span>
  <span>hyperparam_mutations</span><span>=</span><span>{</span>
    <span>'l1_scale'</span><span>:</span> <span>lambda</span><span>:</span> <span>np</span><span>.</span><span>random</span><span>.</span><span>uniform</span><span>(</span><span>1e-3</span><span>,</span> <span>1e-5</span><span>),</span>
    <span>'learning_rate'</span><span>:</span> <span>[</span><span>1e-2</span><span>,</span> <span>1e-3</span><span>,</span> <span>1e-4</span><span>]</span>
  <span>}</span>
<span>)</span>
<span>tune</span><span>.</span><span>run_experiments</span><span>({</span><span>'population_based_training'</span><span>:</span> <span>train_spec</span><span>},</span> <span>scheduler</span><span>=</span><span>pbt</span><span>)</span>
</code></pre></div></div>
<p>The above example will save, explore and exploit your population every time after <code>_train</code> has been called on your Trainable.
This is because we set <em>‘perturbation_interval’</em> to 1.
Furthermore, both <em>‘l1_scale’</em>, as well as <em>‘learning_rate’</em>, will be perturbed or resampled during explore operations according to the scheme specified.
You can even implement your own exploration function, to learn more see <a href="http://ray.readthedocs.io/en/latest/pbt.html">here</a>.</p>

<p>Because at every mutation step of the population-based training the entire model is saved to disk and restored (and possibly sent to a different machine), ray tune requires a significantly bigger overhead compared to our <a href="http://louiskirsch.com/ai/population-based-training">vanilla tensorflow version</a>.
Therefore you might want to increase <em>pertubation_interval</em> depending on the length of each iteration.</p>

<h2 id="optional-network-morphisms">Optional: Network morphisms</h2>

<p>On the other hand, if your graph needs to be rebuilt anyways, for instance, to make architectural changes based on hyperparameters, the graph reconstruction makes the process quite easy.
To give you an example, let’s say you want to increase the number of units in a layer.
You then could implement a network morphism that keeps your neural network function  identical but changes the number of units by padding your weight and bias variables with zeros.
You will have to redefine your <code>_restore</code> function</p>
<div><div><pre><code><span>def</span> <span>_restore</span><span>(</span><span>self</span><span>,</span> <span>checkpoint_path</span><span>):</span>
  <span>reader</span> <span>=</span> <span>tf</span><span>.</span><span>train</span><span>.</span><span>NewCheckpointReader</span><span>(</span><span>checkpoint_path</span><span>)</span>
  <span>for</span> <span>var</span> <span>in</span> <span>self</span><span>.</span><span>saver</span><span>.</span><span>_var_list</span><span>:</span>
    <span>tensor_name</span> <span>=</span> <span>var</span><span>.</span><span>name</span><span>.</span><span>split</span><span>(</span><span>':'</span><span>)[</span><span>0</span><span>]</span>
    <span>if</span> <span>not</span> <span>reader</span><span>.</span><span>has_tensor</span><span>(</span><span>tensor_name</span><span>):</span>
        <span>continue</span>
    <span>saved_value</span> <span>=</span> <span>reader</span><span>.</span><span>get_tensor</span><span>(</span><span>tensor_name</span><span>)</span>
    <span>resized_value</span> <span>=</span> <span>fit_to_shape</span><span>(</span><span>saved_value</span><span>,</span> <span>var</span><span>.</span><span>shape</span><span>.</span><span>as_list</span><span>())</span>
    <span>var</span><span>.</span><span>load</span><span>(</span><span>resized_value</span><span>,</span> <span>self</span><span>.</span><span>sess</span><span>)</span>
</code></pre></div></div>
<p>where <code>fit_to_shape</code> is defined as</p>
<div><div><pre><code><span>def</span> <span>fit_to_shape</span><span>(</span><span>array</span><span>,</span> <span>target_shape</span><span>):</span>
  <span>source_shape</span> <span>=</span> <span>np</span><span>.</span><span>array</span><span>(</span><span>array</span><span>.</span><span>shape</span><span>)</span>
  <span>target_shape</span> <span>=</span> <span>np</span><span>.</span><span>array</span><span>(</span><span>target_shape</span><span>)</span>

  <span>if</span> <span>len</span><span>(</span><span>target_shape</span><span>)</span> <span>!=</span> <span>len</span><span>(</span><span>source_shape</span><span>):</span>
    <span>raise</span> <span>ValueError</span><span>(</span><span>'Axes must match'</span><span>)</span>

  <span>size_diff</span> <span>=</span> <span>target_shape</span> <span>-</span> <span>source_shape</span>

  <span>if</span> <span>np</span><span>.</span><span>all</span><span>(</span><span>size_diff</span> <span>==</span> <span>0</span><span>):</span>
    <span>return</span> <span>array</span>

  <span>if</span> <span>np</span><span>.</span><span>any</span><span>(</span><span>size_diff</span> <span>&gt;</span> <span>0</span><span>):</span>
    <span>paddings</span> <span>=</span> <span>np</span><span>.</span><span>zeros</span><span>((</span><span>len</span><span>(</span><span>target_shape</span><span>),</span> <span>2</span><span>),</span> <span>dtype</span><span>=</span><span>np</span><span>.</span><span>int32</span><span>)</span>
    <span>paddings</span><span>[:,</span> <span>1</span><span>]</span> <span>=</span> <span>np</span><span>.</span><span>maximum</span><span>(</span><span>size_diff</span><span>,</span> <span>0</span><span>)</span>
    <span>array</span> <span>=</span> <span>np</span><span>.</span><span>pad</span><span>(</span><span>array</span><span>,</span> <span>paddings</span><span>,</span> <span>mode</span><span>=</span><span>'constant'</span><span>)</span>

  <span>if</span> <span>np</span><span>.</span><span>any</span><span>(</span><span>size_diff</span> <span>&lt;</span> <span>0</span><span>):</span>
    <span>slice_desc</span> <span>=</span> <span>[</span><span>slice</span><span>(</span><span>d</span><span>)</span> <span>for</span> <span>d</span> <span>in</span> <span>target_shape</span><span>]</span>
    <span>array</span> <span>=</span> <span>array</span><span>[</span><span>slice_desc</span><span>]</span>

  <span>return</span> <span>array</span>
</code></pre></div></div>
<p>Note that in the case where your number of units is reduced, the above code will not keep the function  identical!
One might, for instance, derive a more intelligent algorithm that removes only the weights and biases that have zero magnitudes.
This might be encouraged by l1-regularization.</p>

<h2 id="waiting-for-results-and-visualizing">Waiting for results and visualizing</h2>

<p>Finally, lean back and let the magic happen.
You will see that ray tune outputs nice statistics along the way</p>
<div><div><pre><code>== Status ==
PopulationBasedTraining: 42 checkpoints, 28 perturbs
Resources used: 3/12 CPUs, 3/3 GPUs
Result logdir: /home/louis/ray_results/population_based_training
PAUSED trials:
 - Experiment_0:	PAUSED [pid=20781], 971 s, 1600 ts, -1.64e+03 rew, 0.935 loss, 0.676 acc
RUNNING trials:
 - Experiment_1:	RUNNING [pid=23121], 1162 s, 1350 ts, -1.68e+03 rew, 0.994 loss, 0.665 acc
 - Experiment_2:	RUNNING [pid=18700], 979 s, 1550 ts, -1.63e+03 rew, 0.988 loss, 0.663 acc
 - Experiment_3:	RUNNING [pid=22593], 990 s, 1550 ts, -1.77e+03 rew, 0.959 loss, 0.671 acc
</code></pre></div></div>

<p>To visualize log data both from ray tune and your own tensorflow summaries use Tensorboard</p>
<div><div><pre><code>tensorboard <span>--logdir</span> ~/ray_results/population_based_training
</code></pre></div></div>

<p>If you want to learn more about ray tune, have a look at the <a href="http://ray.readthedocs.io/en/latest/tune.html">documentation</a> and <a href="https://github.com/ray-project/ray/tree/master/python/ray/tune/examples">examples</a>.</p>

  </div></div><br>
<br>
<br>
<br>
AI|ML|DM|IR|PR<br>
<br>
via Louis Kirsch http://louiskirsch.com<br>
<br>
January 7, 2019 at 04:33PM
